### DISCUSSION TITLE
Independent threading for each application

### DISCUSSION THREAD

**Discussion by kangz12345** [snu-quiqcl/qiwis#62] [Ideas]
For each application, I want to spawn a thread dedicated to the application.
This will protect the main process from crashes in applications.

We should study more about Qt's event loop system.
See also: #4.

**Comment by kangz12345** [comment]
However, it seems not easy to give individual threads for each application, since every GUI-related work must be done in the main GUI thread.
This might be tricky... we need to come up with a way to ensure the system robustness.

**Comment by BECATRUE** [comment]
I tested if offering individual threads for each app is able to be a solution.

First of all, refer to the testing branch, `BECATRUE/test/gui-freezing`.

The situation is as below:
- In `numgen` app, when you click the generator button, it will stop 5 seconds using `sleep()`.
- In `swift`, I just offered individual threads for each app. In each for loop,
```python
thread = QThread()
app.moveToThread(thread)
thread.start()
self._threads[name] = thread
```

In conclusion,
- It sometimes works well! This means that all apps work well even if you click the generator button.
- However, it often dies with an error about signal communication between different threads.

Then, how should we do?

**Reply by BECATRUE** [reply]
@kangz12345

**Reply by kangz12345** [reply]
Hmm interesting...
However, I found that the program aborts when I click the "Clear" button of logger.

**Comment by kangz12345** [comment]
@BECATRUE and I decided not to implement this feature, as it is not a feasible plan.
The major problems when we give individual threads to the apps are:
1. All the GUI-modifying methods must be called indirectly (e.g. via signal connections), since it should be executed in the main GUI thread.
2. There is no simple and intuitive way to open a new dialog window from apps (like in `logger` or `dbmgr`).
3. It is natural to program each app assuming that the code is running on the main thread. Each app can run its own thread whenever necessary.
